Add support for collecting instance IP address

songponssw opened the pull request:
## What does this PR do?
- Fix #3616 
- Adds support for generating an instance `instance_info.json` file.
- This file includes basic instance information: _name, IPv4, and IPv6 addresses._


## Implementation details
### With GuestAgent:
- The raw output from `ip -j a` is embedded as `json_bytes` in the gRPC Events message.
- The GuestAgent listens for this event, extracts the IP information, and writes it to the `instance_info.json` file.

### Without GuestAgent (Plain mode):
- This PR updates the `executeSSH()` function signature to return both the command output and error (instead of just error).
- After completing any requirements setup, HostAgent uses `executeSSH()` to run `ip -j a`, processes the output, and writes the instance info locally.

## Testing
- Setup: Lima upstream master, OS(Ubuntu 22.02)
- templates: almalinux alpine debian fedora opensuse oraclelinux ubuntu
- Test command: `limactl start INSTNAME --network lima:user-v2`
-  **Result**
    - With GuestAgent: All tests passed.
    - Plain Mode: Failed on AlmaLinux and Alpine — instances did not start successfully.

AkihiroSuda commented on pkg/guestagent/guestagent_linux.go:
```diff
@@ -186,6 +188,24 @@ func (a *agent) collectEvent(ctx context.Context, st eventState) (*api.Event, ev
 	ev.Time = timestamppb.Now()
+
+	// Get network interfaces info
+	cmd := exec.Command("ip", "-j", "a")
```
Does this really work with busybox?

jandubois replied:
It works with the Alpine cloud image from `template://alpine`. It does not work with alpine-lima from `template://alpine-iso`, but can easily be made to work by adding the `iproute2` package to the image.

afbjorklund replied:
I just used a shell script for apple container, since it "forgot" to update the /etc/hosts file.

`echo "127.0.0.1 localhost" >/etc/hosts`
`echo "$(/sbin/ip -o -4 addr list eth0 | awk '{print $4}' | cut -d/ -f1) $(hostname)" >>/etc/hosts`

afbjorklund replied:
I thought that Alpine used `ip` from busybox by default, but maybe that is what you are saying above

AkihiroSuda commented on pkg/store/filenames/filenames.go:
```diff
@@ -68,7 +68,8 @@ const (
 
-	Protected = "protected" // empty file; used by `limactl protect`
+	Protected = "protected"          // empty file; used by `limactl protect`
+	InstInfo  = "instance_info.json" // store instance information
```
Either way, I don't think the IP has to be recorded in this persistent file

songponssw replied:
Got it! Maybe adding the IP as an attribute in `store.instance` would be sufficient.

AkihiroSuda commented on pkg/hostagent/hostagent.go:
```diff
@@ -803,3 +826,76 @@ func copyToHost(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local,
+		return err
+	}
+
+	var eth0 events.NetworkInterfaceInfo
```
The IP doesn't seem shown in the CLI either.
What's the intended use case of this PR?

AkihiroSuda replied:
> the interface associated with the default route

For what use case?

songponssw replied:
I meant to say it gives the ~interface~ `IP address` associated with the default route, not the interface. Sorry for the mix-up!

AkihiroSuda replied:
The IP associated with the default route isn't guaranteed to be reachable from the host by the IP

AkihiroSuda replied:
But there might not be a robust way to determine the interface reachable from the host by the IP, so relying on the route metric might be acceptable as a heuristic.

Another heuristic would be to match the interface name (`lima%d`)

songponssw commented:
# Changes – Draft 2

### gRPC
- added `CommandOutput` message to hold output from the client.
- Created dedicated endpoints for GetIPv4 and GetIPv6.

### Hostagent
- agent get ipaddress as the last step in host agent routine and it should all only when the instance spin up.
- Stores IP addresses internally
- Exports IP address through the `Info` strut in hostagent/api

### CLI
- Added `AddressIPv4` and `AddressIPv6` fields
- IP address are now available, e.g.: `limactl ls --format "{{ .Name }} {{ .AddressIPv4 }}"`

### Note;
- IPv6 is currently unavailable; no default route is found.
- gRPC compiled grpc with protoc-compiler 6.30.2 

P.S. I'm trying my best to follow the design patterns found in the existing code. If there are any points that need refactoring, please let me know.

songponssw commented:
# PR Update (draft3)

## Details
- Currently supports only IPv4.
- Hostagent requests the guest for its IP through RPC.
- `IPv4Address` is added to Hostagent Info, so it can be shown when using `curl` on `ha.sock` at `http://$INST/v1/info`.

## Determining the IP Address
- The IP is determined by finding an address that can dial `lima.host.internal`.
- `net.Interfaces()` returns a list of all interfaces on the machine. I’m not sure whether there will be only 2 interfaces in practice (loopback and primary interface). If that is the case, the implementation could be simplified to pick the primary interface from the list.

## Items to Figure Out
**Support IPv6:**
- With the current implementation, we would need an IPv6 address for `host.lima.internal`.
- Otherwise, we could assume it uses the same interfaces as IPv4 or dial another destination such as Google DNS or Cloudflare.

**Support Plain Mode:**
- If we want the same implementation to work for both normal and plain mode, we might need to transfer a small binary to the VM to execute and report the IP via RPC.
- I’m not sure if we can reliably get the IP during VM initialization. This might limit the ability for the client to edit the IP after creation. (I’m not quite sure yet exactly when and how IPs are assigned)
- In plain mode, there is no `lima.host.internal` in `/etc/hosts`. We would need to dial another destination to figure out the default route.

songponssw commented on pkg/guestagent/api/guestservice.proto:
```diff
@@ -11,6 +11,8 @@ service GuestService {
+  rpc GetIPv6(google.protobuf.Empty) returns (CommandOutput);
 }
 
 message Info {
```
I’ll try adding the IP address directly to RPC and remove all CommandOutput wrapping
